Order Split View: Show the first item in the secondary column if possible

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #6381 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
There are some existing issues in the order split view regarding the secondary column:
- When the list is first loaded, nothing is displayed on the secondary column even when the list on the primary column is not empty. It'd be nice to automatically display the first item on the secondary column.
- When applying filters to the list, if there was an item displaying on the secondary column, the column still displays the item even when it no longer exists on the list. It'd be nice to switch to the first item on the updated list or show an empty view if the list is empty.

This PR fixes the above issues with the following changes:
- Add a method to check the currently selected item in the list:
  - If no item is selected yet, attempt to highlight the first item in the list and display it on the secondary column.
  - If there's a selected item but no item can be found at the selected index path, switch to displaying the first item instead.
  - If there's an item at the selected index path, but its ID doesn't match the selected order ID, also switch to displaying the first item instead.
- Update the secondary column to display an empty view if we cannot find an appropriate order to display. 


### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Run the app in an iPad or landscape mode of a large-screened iPhone.
- Navigate to Orders tab, notice that the first item in the order list is automatically displayed on the secondary column.
- Apply some filters to the list, notice that the secondary column is updated to show the first item in the list. If the list is empty, the secondary column displays an empty view.
- Please feel free to run the app with the feature flag `.splitViewInOrdersTab` disabled to make sure that everything still works as before.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/5533851/170672356-a9c719f5-dbf1-4004-b571-d5e95011e307.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add check for selected item after dataset of the order list changes

itsmeichigo committed:
Show empty view in the splitview's secondary column if necessary

itsmeichigo committed:
Remove redundant comment

itsmeichigo committed:
Reset selected order ID and index path if no item is found at the selected index path

jaclync reviewed (commented):
Lots of scenarios to handle! 😮 One issue from testing that seems to be from the change, and another unrelated issue:

1. This I think is from the PR change, maybe at `OrdersRootViewController` L436 when both `selectedIndexPath` and `selectedOrderId` are `nil`. I'm not sure why the order list data is non-empty from search results?

Steps to repro:
- Go to the orders tab
- Apply a filter that results in empty results
- Tap search icon to search for orders where there are non-empty results
- Tap on any order in the search results --> on iPad, notice that the detail screen in the background shows the selected order
- Dismiss the search screen --> the detail screen still shows the selected order from search

https://user-images.githubusercontent.com/1945542/170912089-38b9ae6f-0458-4552-b273-adc891e039b9.mov


2. This is unrelated to the PR, just wanted to mention: after creating an order, it's pushed to the first column instead being shown in the secondary column:

<img src="https://user-images.githubusercontent.com/1945542/170912425-67c1e19b-1523-4e2a-8bca-ff45c32dfc4c.png" width="300" />

itsmeichigo commented:
Thank you @jaclync for the review and for noticing those issues!

For the first issue above, it seems that loading order details from the search flow triggers updates in storage, which in turn triggers the snapshot update in the order list with that single item. I'm not sure if this should be fixed by stopping receiving snapshots when the search screen is opened - this sounds more like a workaround and doesn't look nice.

A similar issue I've found is that we're also configuring the order details screen to [listen to updates of Order entities](https://github.com/woocommerce/woocommerce-ios/blob/2c5585f22b174f52361bcb732b491df10785d006/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderDetailsViewController.swift#L147-L155) - which means whenever an item is selected on the search flow, you can notice that the item on the secondary column of the split view is updated as well. This issue is more minor though.

I haven't come up with a good solution for these, so I'll convert this PR back to draft until I can get back to it.

> This is unrelated to the PR, just wanted to mention: after creating an order, it's pushed to the first column instead being shown in the secondary column:

I've also logged this as an item in #6381 to fix later.

jaclync commented:
It looks like we [set a boolean flag `exclusiveForSearch` to `Order` when inserting a new order in search results](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift#L21-L27), and the [order list does exclude orders with `exclusiveForSearch = false`](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L197). I wonder if something was broken in setting this boolean flag 🤔 

> A similar issue I've found is that we're also configuring the order details screen to [listen to updates of Order entities](https://github.com/woocommerce/woocommerce-ios/blob/2c5585f22b174f52361bcb732b491df10785d006/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderDetailsViewController.swift#L147-L155) - which means whenever an item is selected on the search flow, you can notice that the item on the secondary column of the split view is updated as well. This issue is more minor though.

I felt that we should guard that the updated order matches the order ID in order details in the order listener?

itsmeichigo commented:
> It looks like we [set a boolean flag exclusiveForSearch to Order when inserting a new order in search results](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift#L21-L27), and the [order list does exclude orders with exclusiveForSearch = false](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift#L197). I wonder if something was broken in setting this boolean flag 🤔

I don't think this is the cause - the issue happens only when selecting an order from the search result. So when opening the order details screen, we're calling [`syncOrder`](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L340-L352), which will trigger retrieving order and then upsert the order into the local storage. This will then trigger the snapshot update in the order list.

> I felt that we should guard that the updated order matches the order ID in order details in the order listener?

Let's scratch what I said about the entity listener, it seems like the [listener to payment gateway account changes](https://github.com/woocommerce/woocommerce-ios/blob/eb36ab429f5876c8d584989bf6784d5e396796e9/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsResultsControllers.swift#L261-L263) that triggers the update. We're fetching this information again upon opening any order - I'm not sure if it's really necessary? How about doing that only when opening the Orders tab and the Order search screen?

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-orders-secondary-column

itsmeichigo committed:
Do not select first item automatically if the list is empty

itsmeichigo committed:
Do not automatically select first item if list's state is empty

itsmeichigo commented:
I've revisited this PR today and found that the issue with the search result list can be easily fixed by making sure that the order list is not empty before automatically selecting the first item. I fixed this in https://github.com/woocommerce/woocommerce-ios/pull/6972/commits/27a4d3940d30a198b537cb4a921450c17ee94a47, so @jaclync could you please take this another look? 


https://user-images.githubusercontent.com/5533851/175272441-dece037d-3fd0-4319-8ae2-4ed7109b1ad6.mp4

jaclync commented:
Hey @itsmeichigo! I just tested the latest commit, and saw these unexpected behaviors on the first load of the orders tab on iPhone and iPad:

- On iPad, the "No order selected" screen has a back button that navigates back to the same screen

https://user-images.githubusercontent.com/1945542/175356999-9d773eda-39c6-4289-a427-11d5b17c5042.mov

- On iPhone, the orders tab shows the first order in the list

https://user-images.githubusercontent.com/1945542/175357410-db128d08-baf9-4714-8212-15dd69fbafa5.mov

The second one feels like it might be from the PR changes?

itsmeichigo committed:
Only automatically select first item if the split view is displaying two columns

itsmeichigo committed:
Embed the empty state view controller in a navigation controller before displaying it in split view

itsmeichigo commented:
@jaclync Thanks again for pointing out those issues! I've made another attempt to fix them and hopefully it works properly now:

> On iPad, the "No order selected" screen has a back button that navigates back to the same screen

This is because I'm showing the empty view without a navigation controller, and it gets pushed into an existing navigation controller. I fixed this in 3cd206145384189e1bc42628f41f75905ce97140 by embedding it inside a navigation controller, this way it's always the root item when shown from the split view.

> On iPhone, the orders tab shows the first order in the list

This is an obvious issue that I missed because I didn't test on the iPhone 😅. I fixed it by making sure that the split view is not collapsed (i.e showing 2 columns) before doing the automatic selection in 859d19e9e3421a277e7759ec69d617acd7ad4181.

Please take another look and let me know if everything is good now, thanks!

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -103,12 +103,16 @@ final class OrderListViewController: UIViewController, GhostableViewController {
 
+    /// Currently selected order ID in the table view
+    ///
+    private var selectedOrderId: Int64?
```
super nit:
```suggestion
    private var selectedOrderID: Int64?
```

jaclync reviewed (approved):
Nice work handling so many scenarios @itsmeichigo! 🎉 🚢

itsmeichigo committed:
Update WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift

Co-authored-by: Jaclyn Chen <watertranquil@gmail.com>

itsmeichigo committed:
Rename selectedOrderId to selectedOrderID
